h2o-clustering/src/main/java/water/clustering/AssistedClusteringEmbeddedConfigProvider.java

```diff
@@ -0,0 +1,79 @@
+package water.clustering;
+
+import org.apache.log4j.Logger;
+import water.clustering.api.AssistedClusteringRestApi;
+import water.clustering.api.GracefulAsyncRunner;
+import water.init.AbstractEmbeddedH2OConfig;
+import water.init.EmbeddedConfigProvider;
+import water.util.Log;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Provides {@link AssistedClusteringEmbeddedConfig} as a result of external assist. The resulting
+ * embedded config provides a flatfile assembled by external service, given to H2O. 
+ * This EmbeddedConfig has no timeout and will wait indefinitely until a flatfifle is received.
+ */
+public class AssistedClusteringEmbeddedConfigProvider implements EmbeddedConfigProvider {
+    private static final Logger LOG = Logger.getLogger(AssistedClusteringEmbeddedConfigProvider.class);
+
+    private AssistedClusteringEmbeddedConfig assistedClusteringEmbeddedConfig;
+    private String flatfile;
+    private final Object notification = new Object();
+
+    @Override
+    public void init() {
+        final Consumer<String> flatFileCallback = s -> {
+            // Make sure to set the flatfile first and then notify, as the flatfile is immediately used
+            // to produce a resulting config
+            flatfile = s;
+            synchronized (notification) {
+                notification.notify();
+            }
+        };
```

Comment:
I think the consumer-producer code could be simplified by using SynchronousQueue to exchange the flatfile. The advantage is that you can use a timeout on the consumer side and you can occasionally produce logs about how the process is going. I would use XGBoostUpdater#invoke for inspiration.

Reply:
The timeout is something I'm not sure we want.

Reply:
Done. I kept the callback interface inside the REST API to make it more generic and just did the queue on the client of the endpoint. Also made the callback asynchronous, so it's not blocking the REST API.